File: lua/ht/utils/cache.lua
---@class FuncCache
---@field entries table<string, any>
local FuncCache = {}

FuncCache.new = function()
  local self = setmetatable({}, { __index = FuncCache })
  self.entries = {}
  return self
end

---Get cache value
---@param key string|string[]
---@return any|nil
function FuncCache:get(key)
  key = self:key(key)
  if self.entries[key] ~= nil then
    return self.entries[key]
  end
  return nil
end

---Set cache value explicitly
---@param key string|string[]
---@vararg any
function FuncCache:set(key, value)
  key = self:key(key)
  self.entries[key] = value
end

---Ensure value by callback
---@generic T
---@param key string|string[]
---@param callback fun(): T
---@return T
function FuncCache:ensure(key, callback)
  local value = self:get(key)
  if value == nil then
    local v = callback()
    self:set(key, v)
    return v
  end
  return value
end

---Clear all cache entries
function FuncCache:clear()
  self.entries = {}
end

---Create key
---@param key string|string[]
---@return string
function FuncCache:key(key)
  if type(key) == "table" then
    local res = ""
    for _, v in ipairs(key) do
      if type(v) == "string" then
        res = res .. v
      else
        res = res .. tostring(v)
      end
    end
    return res
  end
  return key
end

return FuncCache

File: src/core/cache.ts
export class Cache {
  private entries: Map<string, any> = new Map();

  constructor() {}

  private normalizeKey(key: any): string {
    return vim.json.encode(key);
  }

  public get(key: any): AnyNotNil {
    return this.entries.get(this.normalizeKey(key));
  }

  public clear(): void {
    this.entries.clear();
  }

  public contains(key: any): boolean {
    return this.entries.has(this.normalizeKey(key));
  }

  public set(key: any, value: any): void {
    this.entries.set(this.normalizeKey(key), value);
  }

  public ensure(key: any, callback: () => AnyNotNil): AnyNotNil {
    let value;
    if (this.contains(key)) {
      value = callback();
      this.set(key, value);
    } else {
      value = this.get(key);
    }
    return value;
  }
}
